cores/arduino/stm32/usb/usbd_if.c

```diff
@@ -10,33 +10,77 @@
+#if !defined(USBD_ATTACH_PINNAME) && !defined(USBD_DETACH_PINNAME) && !defined(USB_OTG_DCTL_SDIS)
+#if defined(USE_USB_HS_IN_FS)
+#define USBD_DETACH_PINNAME USB_OTG_HS_DP
+#elif defined(USB_OTG_FS)
+#define USBD_DETACH_PINNAME USB_OTG_FS_DP
+#else /* USB */
+#define USBD_DETACH_PINNAME  USB_DP
+#endif
+#define USBD_DETACH_LEVEL LOW
+// Make sure to revert to INPUT (instead of OUTPUT HIGH) after the
+// detach "pulse" to restore normal operation.
+#define USBD_ATTACH_MODE INPUT
+#endif /* !defined(USBD_ATTACH_PINNAME) && !defined(USBD_DETACH_PINNAME) && !defined(USB_OTG_DCTL_SDIS) */
+
+#if defined(USBD_DETACH_PINNAME) && !defined(USBD_ATTACH_MODE)
+// Normally (e.g. when controlling a transistor), keep the pin in
+// OUTPUT mode after the detach "pulse" to actively drive the
+// transistor.
+#define USBD_ATTACH_MODE OUTPUT
+#endif /* defined(USBD_ATTACH_PINNAME) && !defined(USBD_ATTACH_MODE) */
+
+// Some sanity checks
+#if defined(USBD_ATTACH_PINNAME) && defined(USBD_DETACH_PINNAME)
+#error "Cannot define both USBD_ATTACH_PINNAME and USBD_DETACH_PINNAME"
+#endif /* defined(USBD_ATTACH_PINNAME) && defined(USBD_DETACH_PINNAME) */
+#if defined(USBD_ATTACH_PINNAME) && !defined(USBD_ATTACH_LEVEL)
+#error "USBD_ATTACH_PINNAME also needs USBD_ATTACH_LEVEL defined"
+#endif /* defined(USBD_ATTACH_PINNAME) && !defined(USBD_ATTACH_LEVEL) */
+#if defined(USBD_DETACH_PINNAME) && !defined(USBD_DETACH_LEVEL)
+#error "USBD_DETACH_PINNAME also needs USBD_DETACH_LEVEL defined"
+#endif /* defined(USBD_DETACH_PINNAME) && !defined(USBD_DETACH_LEVEL) */
+
 /**
   * @brief  Force to re-enumerate USB
   * @param  None
   * @retval None
   */
-
 WEAK void USBD_reenumerate(void)
 {
 #ifndef USBD_REENUM_DISABLED
   /* Re-enumerate the USB */
-#ifdef USB_DISC_PIN
-  pinMode(USB_DISC_PIN, OUTPUT);
-  digitalWrite(USB_DISC_PIN, LOW);
-#else
-#ifdef USE_USB_HS_IN_FS
-  PinName pinDP = USB_OTG_HS_DP;
-#elif defined(USB_OTG_FS)
-  PinName pinDP = USB_OTG_FS_DP;
-#else /* USB */
-  PinName pinDP =  USB_DP;
-#endif
-  pin_function(pinDP, STM_PIN_DATA(STM_MODE_OUTPUT_PP, GPIO_NOPULL, 0));
-  digitalWriteFast(pinDP, LOW);
+#if defined(USBD_DETACH_PINNAME)
+  // Default is attached, so detach, wait, and reattach.
+  pin_function(USBD_DETACH_PINNAME, STM_PIN_DATA(STM_MODE_OUTPUT_PP, GPIO_NOPULL, 0));
+  digitalWriteFast(USBD_DETACH_PINNAME, USBD_DETACH_LEVEL);
   delay(USBD_ENUM_DELAY);
-  pin_function(pinDP, STM_PIN_DATA(STM_MODE_INPUT, GPIO_NOPULL, 0));
-  /*delay(USBD_ENUM_DELAY);*/
-#endif /* USB_DISC_PIN */
+  if (USBD_ATTACH_MODE == OUTPUT) {
+    digitalWriteFast(USBD_DETACH_PINNAME, !USBD_DETACH_LEVEL);
+  } else {
+    pin_function(USBD_DETACH_PINNAME, STM_PIN_DATA(STM_MODE_INPUT, GPIO_NOPULL, 0));
+  }
+#elif defined(USBD_ATTACH_PINNAME)
+  // Default is detached, so just attach
+  pin_function(USBD_ATTACH_PINNAME, STM_PIN_DATA(STM_MODE_OUTPUT_PP, GPIO_NOPULL, 0));
+  digitalWriteFast(USBD_ATTACH_PINNAME, USBD_ATTACH_LEVEL);
+#elif !defined(USB_OTG_DCTL_SDIS)
```

Comment:
I didn't catch why we do not set then reset bit USB_OTG_DCTL_SDIS in case of Internal PullUp ? this should force reenumeration, isn't it ? otherwise, how reenumeration is performed ?

Reply:
The pullup is automatically controlled by hardare, so it is already inactive during reset and activated once the USB core is initialized, so nothing is needed here during startup (SDIS won't actually *do* anything when the USB core is not initialized).

One case where SDIS could be useful is to re-attach during normal operation. This is the third usecase I described in #885, and though I planned to support (I see now), this PR does not implement it yet. Might be good to add anyway, which would involve triggering toggling SDIS, and also disabling the ATTACH pin (which is now only enabled, never disabled).

I'm not sure how the USB core handles a random detach / attach, but I guess it won't be much different from actually replugging USB while externally powered, which the code probably handles? I'll test :-)